File: testDataOutput/19_interfaces.swift
import Foundation
// Simple Interface
//package com.lightningkite.interfaceTest

//import com.lightningkite.others.*

public protocol MyInterface {
    func bar() -> String
    var x: Int
  { get set }    var y: String
  { get }    var z: Float { get { return 0f }
 }
}

final public class Implementation : MyInterface {

    init() {

    }
    public func bar() -> String {
        return "2-Implementation"
    }

    public var x: Int {
        get { return 2 }
        set(value) {}
        }
    public var y: String { get { return "Hello World!" }
 }
}

// Interface + Inheritance

public open class Parent {

    init() {

    }
    func three() -> Int {
        return 3
    }

    var four = 4 {
        didSet(value) {
            
            print(value)
        }
        }
}

final public class Child : Parent(), MyInterface {

    init() {

        super.init()
    }
    var five = 5
    func six() -> String {
        return "6"
    }

    public func bar() -> String {
        return "2-Child"
    }
}

// Abstract Interface + Inheritance

public abstract class AbstractParent {

    init() {

    }
    abstract func three() -> Int

    var four = 4
}

public abstract class AbstractChild : Parent() {

    init() {

        super.init()
    }
}

final public class NonAbstractChild : AbstractChild(), MyInterface {

    init() {

        super.init()
    }
    public func bar() -> String {
        return "2-NonAbstractChild"
    }
}

func main(args: [String]) {
    // Simple Interface
    var obj = Implementation()
    print(obj.foo())
    print(obj.bar())

    // Interface + Inheritance
    var child = Child()
    print(child.foo())
    print(child.bar())
    print(child.three())
    print(child.four)
    print(child.five)
    print(child.six())

    // Abstract Interface + Inheritance
    var naChild = NonAbstractChild()
    print(naChild.foo())
    print(naChild.bar())
    print(naChild.three())
    print(naChild.four)
}

File: testData/shared/19_interfaces.kt
// Simple Interface
package com.lightningkite.interfaceTest

import com.lightningkite.others.*

interface MyInterface {
    fun bar(): String
    var x: Int
    val y: String
    val z: Float get() = 0f
}

class Implementation : MyInterface {
    override fun bar(): String {
        return "2-Implementation"
    }

    override var x: Int
        get() = 2
        set(value) {}
    override val y: String get() = "Hello World!"
}

// Interface + Inheritance

open class Parent {
    fun three(): Int {
        return 3
    }

    var four = 4
        set(value) {
            field = value
            println(value)
        }
}

class Child : Parent(), MyInterface {
    val five = 5
    fun six(): String {
        return "6"
    }

    override fun bar(): String {
        return "2-Child"
    }
}

// Abstract Interface + Inheritance

abstract class AbstractParent {
    abstract fun three(): Int
    var four = 4
}

abstract class AbstractChild : Parent() {
}

class NonAbstractChild : AbstractChild(), MyInterface {
    override fun bar(): String {
        return "2-NonAbstractChild"
    }

    override fun three(): Int = 3
}

fun main(args: Array<String>) {
    // Simple Interface
    val obj = Implementation()
    println(obj.foo())
    println(obj.bar())

    // Interface + Inheritance
    val child = Child()
    println(child.foo())
    println(child.bar())
    println(child.three())
    println(child.four)
    println(child.five)
    println(child.six())

    // Abstract Interface + Inheritance
    val naChild = NonAbstractChild()
    println(naChild.foo())
    println(naChild.bar())
    println(naChild.three())
    println(naChild.four)
}
